src/inflators/model.tsx

```diff
@@ -88,10 +88,6 @@ export function inflateModel(world: HubsWorld, rootEid: number, { model }: Model
 
     const replacement = world.eid2obj.get(eid);
     if (replacement) {
-      if (obj.type !== "Object3D") {
-        console.error(obj, replacement);
-        throw new Error("Failed to inflate model. Unexpected object type found before swap.");
-      }
```

Comment:
I first want to know the original purpose of this check...

It looks like indicating that if a Hubs component generates Three.js Object, the glTF node that has the Hubs component shouldn't be Mesh (or other object) node. Do we have such a limitation in Hubs components?

```
// glTF
"nodes": [
  {
    // Valid
    "extensions": {
      "MOZ_hubs_components": {
        // foo generates Three.js Object3D in Hubs
        "foo": ...
      }
    }
  },
  {
    // Invalid
    "mesh": 0,
    "extensions": {
      "MOZ_hubs_components": {
        "foo": ...
      }
    }
  }
]
```

Reply:
My interpretation of this is that for some reason we don't allow components that create meshes to replace a mesh but I'm not sure why that limitation and more importantly this is not how the old loader works and that can break existing scenes.

This patch won't make us assign multiple objects to a single entity if I'm correct, this just allows replacing meshes if the have a component attached that creates a mesh.

Maybe @johnshaughnessy has more info about why it's important that we don't override meshes (again, my main concern if scene brekeage as this is not how the AFrame loader works)

Reply:
>  just allows replacing meshes

If I understand correctly...

glTF mesh in a node will be ignored if Hubs components that generate Three.js objects are set in the same node, if we allow replace meshes.

```
// glTF
"nodes": [
  {
    "extensions": {
      "MOZ_hubs_components": {
        // foo generates Three.js Object3D in Hubs
        "foo": ...
      }
    },
    // this mesh will be ignored because the mesh object will be
    // replaced with objects generated from `foo` component
    mesh: 0
  },
]
```

Is it an expected behavior? For me, it would be a weird behavior tho.

Perhaps we may need to learn the history from @johnshaughnessy 

If my assumption is correct and this behavior is weird but we want to add a hack for old loader compatilibity, we may need to add a comment describing that this is a hack and should be removed when we get rid of the old compatibility.

Reply:
This limitation is intentional. We should not remove it.

In the new code, each `entity` in hubs is associated with _at most_ one `Object3D` in the scene graph. This restriction simplifies the code in a number of places. 

Our old code would insert `Group` nodes into the scene graph whenever the components on a node called for multiple `Three.js` objects. These changes to the scene graph added a performance cost and increased complexity: We needed to add code to rename nodes, retarget animations, and copy `morphTargetInfluences`.

We want each component added to a node in a gltf file to unambiguously target an entity when loaded into Hubs. Suppose we allow a node to have a mesh and an audio component and a text component, and then we add a `Foo` component. To which `entity`/`object3D` should we assign the `Foo`? The `mesh`, the `audio`, the `text`, or a `Group` node containing them? We decided that there is no _good_ answer to this question in general, and we should _reject_ these gltf files.

As you have rightly pointed out, though, there are lots of gltf files that already break this rules. We should solve this "at the edge", rather than here "in the middle" of the code. Specifically, we should try to transform an invalid `gltf` into a valid one within `GLTFLoader`.

You can see an example of how we already "patch" gltf files before loading them here: https://github.com/mozilla/hubs/blob/31373a2c660e6a349a5548ff5de31bcdf1c9fce6/src/components/gltf-model-plus.js#L508-L521

I think we should similarly try to introduce whatever changes are necessary to a gltf at this earliest possible stage, and then throw informative warnings to the user to let them know what's wrong. Similarly, our hubs blender exporter should warn on these issues and tell the author how to fix them.

In order to keep the client as simple as possible, we need a simple scene graph. We should not enable gltf authors to create a `Light` that is also an `Audio` that is also a `Mesh`, because we cannot represent such a thing in Three. If they want an entity with three _children_ (a light, an audio, and a mesh), that's OK. But they should explicitly create that structure.

Reply:
Perhaps we may also add validation when exporting (with Hubs Blender add-on or from Spoke).

Reply:
Thanks for the background on this decision, it makes sense. As you said we can add some warnings on the Blender side (Spoke doesn't have this issue) when adding a "mesh" component and in the migration report, that is something that could help the user to be aware of this limitation and also help them understand which scenes require updates. I'm going to close this PR and add it to the add-on for the next release.